File: src/leetCode/c++/Solution763.cpp
//Greedy Slide Window O(n) O(Z), Z = 26
class Solution763 {
public:
    vector<int> partitionLabels(string s) {
        unordered_map<char, int> m;//<字母, 最後出現位置>
        for(int i=0; i<s.length(); ++i)
            m[s[i]] = i;

        vector<int> res;
        int l = 0, r = 0;
        for(int i=0; i<s.length(); ++i){
            r = max(r, m[s[i]]);
            if(i == r){//此時結束一個段落
                res.push_back(r - l + 1);
                l = r + 1;
            }
        }
        return res;
    }
};
/* 只需要關心該字母最後出現的位置即可

abcbaefgf
  ^^^^ ^^
*/

File: java/Solution763.java
package leetCode.java;

import java.util.*;

//Greedy Slide Window O(n) O(Z), Z = 26
class Solution763 {
    public List<Integer> partitionLabels(String s) {
        int[] lastIndex = new int[26]; // 該字母最後出現的位置
        for (int i = 0; i < s.length(); ++i) {
            lastIndex[s.charAt(i) - 'a'] = i;
        }

        int l = 0, r = 0;
        List<Integer> res = new ArrayList<>();
        for (int i = 0; i < s.length(); ++i) {
            l = Math.max(l, lastIndex[s.charAt(i) - 'a']);
            if (l == i) { // 走到斷點, 完成一個段落
                res.add(l - r + 1);
                r = l + 1;
            }
        }
        return res;
    }
}
/*
a b c b a d e f f g e
--------- - ---------
  -----       --- -
    -

=>[5,1,5]

*/